(chore): add e2e tests for encryption-at-rest

buraksekili opened the pull request:
**What this PR does / why we need it**:

This commit adds e2e tests for the encryption-at-rest functionality in KKP. The changes include:

- Add new e2e test package for encryption-at-rest with test cases for enabling and disabling encryption
- Create a CI script to run the encryption-at-rest e2e tests
- Update Prow configuration to add encryption-at-rest e2e tests to the CI pipeline
- Fix parameter handling in the encryption-at-rest controller
- Add verification for proper encryption and decryption of Kubernetes secrets

The tests validate that:
1. Enabling encryption works correctly and data gets encrypted
2. Disabling encryption works correctly and data gets decrypted
3. API server configuration is properly updated with encryption settings

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #8405

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

buraksekili committed:
feat: add e2e tests for encryption-at-rest feature

This commit adds comprehensive end-to-end tests for the encryption-at-rest functionality in Kubermatic Kubernetes Platform. The changes include:

- Add new e2e test package for encryption-at-rest with test cases for enabling and disabling encryption
- Create a CI script to run the encryption-at-rest e2e tests
- Update Prow configuration to add encryption-at-rest e2e tests to the CI pipeline
- Fix parameter handling in the encryption-at-rest controller
- Add verification for proper encryption and decryption of Kubernetes secrets

The tests validate that:
1. Enabling encryption works correctly and data gets encrypted
2. Disabling encryption works correctly and data gets decrypted
3. API server configuration is properly updated with encryption settings

embik commented on .prow/features.yaml:
```diff
@@ -308,3 +308,39 @@ presubmits:
               memory: 6Gi
+
+  - name: pre-kubermatic-encryption-at-rest-e2e
+    run_if_changed: "(pkg/controller/seed-controller-manager/encryption-at-rest-controller/|pkg/test/e2e/encryption-at-rest/)"
```
I'm wondering if this should be run more frequently than with those files, e.g. when we make changes to the Kubernetes version or the apiserver.

buraksekili replied:
Makes sense! Do you think we should run this test case against every PR?

embik replied:
Every commit might be a bit much, but I think it makes sense to test it against a) any apiserver changes and b) any Kubernetes version changes (which is somewhere in `pkg/defaulting`, afaik?).

buraksekili replied:
Sounds good! I’m planning to run this test case if there are changes in `pkg/defaulting` or `pkg/resources`. Does that approach work for you?

embik replied:
I think `pkg/resources/apiserver` should be sufficient, otherwise lgtm!

buraksekili replied:
Thanks for the review. Updated accordingly in 238d6da05ff1a1f2b0a28bb605a3200635034771

embik replied:
Sorry, I wasn't clear in what I meant to say - your previous triggers were perfect, the addition of `pkg/defaulting` makes sense, I just meant to say instead of `pkg/resources` you can "just" add `pkg/resources/apiserver`. Not replace everything with it, but I can see how it could be read that way.

buraksekili replied:
oh, my bad. i misunderstood, thanks for the clarification :) 
i've updated the PR accordingly.

embik commented on pkg/test/e2e/encryption-at-rest/ear_test.go:
```diff
@@ -0,0 +1,464 @@
+				return fmt.Errorf("failed to get cluster: %w", err), nil
+			}
+
+			r.logger.Infof("cluster status: %+v", c.Status)
```
I think this might be a bit much, maybe it's a leftover from development debugging?

embik commented on pkg/test/e2e/encryption-at-rest/ear_test.go:
```diff
@@ -0,0 +1,464 @@
+				return fmt.Errorf("failed to get cluster: %w", err), nil
+			}
+
+			r.logger.Infof("cluster status: %+v", c.Status)
```
same here.

buraksekili replied:
thanks! removed these logs in 19839f8a79cc82b7d06b8a34d23949a35a2deb1f

embik reviewed (commented):
Some minor nits.

buraksekili committed:
Merge branch 'main' into chore/etcd-enc-e2e

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
run encryption at rest test cases when pkg/resources has changed

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
Merge branch 'main' into chore/etcd-enc-e2e

buraksekili committed:
fix linter issues

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
update the condition that controls when to run encryption at rest e2e tests

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
Merge branch 'main' into chore/etcd-enc-e2e

buraksekili committed:
feat(encryption-at-rest): enhance e2e tests with backup/restore validation

- Implementing key rotation functionality for encryption keys
- Moved BYO cluster from AWS
- Adding backup and restore test cases to verify data encryption with different keys
- Testing data accessibility after etcd restore operations
- Updating test scripts to use backup-enabled environment setup

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

move BYO cluster from AWS

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

add complete object reference of the cluster and add debug statements

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

enable etcd launcher for etcd backups

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

enable etcd launcher for etcd backups

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

fix linter

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

ensure TLS peering

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

fix timeout

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

decrease the number of characters in cluster name to align with k8s constraints

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

use correct cluster reference

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

wait for etcd to restore

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

update key rotation flow

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

fix key used after rotation

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

update key

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

embik commented on pkg/test/e2e/encryption-at-rest/ear_test.go:
```diff
@@ -0,0 +1,906 @@
+
+	// Test Case 1:
+	// Enable encryption at rest and verify it works
+	logger.Info("Test Case 1 running...")
```
A minor style note, given this is running end-to-end as what I would call a single test case (we aren't doing table testing or anything like that), I would suggest changing the log lines for this (and all the other "Test Case x running"). The comments above each lines are explaining well what is being tested, maybe we can put that information (in abbreviated form) into the logs?

Something like

```suggestion
	logger.Info("Enabling encryption at rest...")
```

And similar for later log lines.

buraksekili replied:
Thanks! removed redundant log statements in 6379af519a1da088688f2a62047439ad73d818c5. let me know if it needs further work

buraksekili committed:
simplify the logs, remove redundant log statements for brevity

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

embik reviewed (approved):
/approve

🎉
